File: studio/screens/room.js
// Phase 4 — Room detail page. One page per real zone from core/twin.js's
// deriveZones() (living / utility / panel — the only zones this system
// actually has data for). Shows real points, the real action/incident
// timeline where one genuinely applies to the zone, and health. No 3D view,
// predictions, automation, or maintenance suggestions — none of that is
// computed anywhere in this codebase yet; see docs/APP_PLAN.md §7.
import { h, mount, raf } from "../core/dom.js";
import { store } from "../core/store.js";
import { deriveZones, deriveWater, deriveOccupancy, derivePower } from "../core/twin.js";
import { navigate } from "../core/router.js";
import { fmt, STATE } from "../core/format.js";
import { healthDot, chip } from "../ui/health-dot.js";
import { lineChart } from "../ui/line-chart.js";

const ZONES = {
  living: {
    label: "Living",
    pointKeys: ["house.occupied", "living.pir", "living.radar"],
    chartKey: null,
    actionsRelevant: false,
  },
  utility: {
    label: "Water / Utility",
    pointKeys: [
      "tank.line.flow_lpm", "water_tank.level_pct", "main_valve.valve_state",
      "pump.pump_state", "pump.current_a", "virtual.water.leak_suspected",
      "virtual.pump.dryrun_suspected", "health.main_valve", "health.pump",
    ],
    chartKey: "water_tank.level_pct",
    actionsRelevant: true,
  },
  panel: {
    label: "Main Panel",
    pointKeys: ["main_panel.power_w"],
    chartKey: "main_panel.power_w",
    actionsRelevant: false,
  },
};

let unsubscribe = null;

export function renderRoom(target, params) {
  if (unsubscribe) unsubscribe();
  const paint = raf(() => mount(target, view(store.state, params.id)));
  paint();
  unsubscribe = store.subscribe(paint);
}

function view(state, id) {
  const zoneDef = ZONES[id];
  if (!zoneDef) {
    return h("div", { class: "glass card" }, [
      h("h2", {}, "Unknown room"),
      h("div", { class: "empty" }, `No zone named "${id}". Try Living, Utility, or Panel from the Home screen.`),
    ]);
  }

  const zones = deriveZones(state.points);
  const zone = zones.find((z) => zoneDef.label === z.label) || { state: STATE.OK, detail: "—" };
  const water = deriveWater(state.points);
  const occ = deriveOccupancy(state.points);
  const power = derivePower(state.points);

  return h("div", { class: "room-screen" }, [
    h("div", { class: `glass card room-header ${zone.state}` }, [
      healthDot(zone.state),
      h("div", {}, [
        h("h1", { class: "room-title" }, zoneDef.label),
        h("div", { class: "room-detail" }, zone.detail),
      ]),
    ]),

    zoneDef.chartKey ? h("div", { class: "glass card" }, [
      h("h2", {}, "Trend — this session"),
      lineChart({ data: store.series(zoneDef.chartKey), valueFmt: (v) => v.toFixed(1) }),
    ]) : null,

    id === "living" ? h("div", { class: "glass card" }, [
      h("h2", {}, "Occupancy"),
      h("div", { class: "occ-row" }, [
        chip(occ.occupied ? "occupied" : "empty", occ.occupied ? STATE.OK : STATE.LEARN),
        h("span", { class: "room-mini" }, `radar ${occ.radar ? "on" : "off"} · PIR ${occ.pir ? "on" : "off"}`),
      ]),
    ]) : null,

    id === "utility" ? h("div", { class: "water-status-row" }, [
      chip(water.leak ? "leak suspected" : "no leak", water.leak ? STATE.CRIT : STATE.OK),
      chip(water.dryrun ? "pump dry-run" : "pump nominal", water.dryrun ? STATE.CRIT : STATE.OK),
    ]) : null,

    id === "panel" && !Number.isNaN(power.totalW)
      ? h("div", { class: "glass card" }, [h("h2", {}, "Power now"), h("div", { class: "room-power" }, `${power.totalW.toFixed(0)} W`)])
      : null,

    h("div", { class: "glass card" }, [
      h("h2", {}, "Points in this room"),
      pointsList(state, zoneDef.pointKeys),
    ]),

    zoneDef.actionsRelevant ? h("div", { class: "glass card" }, [
      h("h2", {}, "Autonomous actions in this room"),
      state.actionOrder.length
        ? h("div", { class: "incident-list" }, state.actionOrder.map((aid) => actionRow(state.actions.get(aid))))
        : h("div", { class: "empty" }, "No autonomous action yet in this run."),
    ]) : h("div", { class: "glass card" }, [
      h("h2", {}, "Autonomous actions"),
      h("div", { class: "empty" }, "This system never acts on this room — only the water loop (utility) is autonomous today."),
    ]),
  ]);
}

function pointsList(state, keys) {
  const rows = keys.map((key) => {
    const p = state.points.get(key);
    return h("div", { class: "sensor-row", onclick: () => navigate(`/sensor/${encodeURIComponent(key)}`) }, [
      h("span", { class: "sensor-key" }, key),
      h("span", { class: "sensor-value" }, p ? fmt(p.value) : "—"),
      h("span", { class: "sensor-arrow" }, "›"),
    ]);
  });
  return h("div", { class: "sensor-list" }, rows);
}

function actionRow(a) {
  const statusChip = a.status === "confirmed" ? chip("verified", STATE.OK)
    : a.status === "failed" ? chip("escalated", STATE.CRIT)
    : chip("pending", STATE.LEARN);
  return h("div", { class: "incident-row" }, [
    h("div", { class: "incident-head" }, [
      healthDot(a.status === "confirmed" ? STATE.OK : a.status === "failed" ? STATE.CRIT : STATE.LEARN),
      h("span", { class: "incident-cause" }, a.cause || `action #${a.id}`),
      statusChip,
    ]),
  ]);
}

File: mobile/lib/screens/room.dart
// Phase D — Room detail. Port of studio/screens/room.js. One page per real
// zone from core/twin.dart's deriveZones() (living / utility / panel — the
// only zones this system actually has data for). Real points, real health,
// and the real action timeline where one genuinely applies to the zone. No 3D
// view, predictions, automation, or maintenance suggestions: none of that is
// computed anywhere in this codebase. See docs/APP_PLAN.md §7.
import 'package:flutter/material.dart';

import '../core/app_scope.dart';
import '../core/format.dart';
import '../core/store.dart';
import '../core/twin.dart';
import '../theme/tokens.dart';
import '../widgets/glass_card.dart';
import '../widgets/health_dot.dart';
import '../widgets/incident_row.dart';
import '../widgets/line_chart.dart';
import 'sensor.dart';

class _ZoneDef {
  final String label;
  final List<String> pointKeys;
  final String? chartKey;
  final bool actionsRelevant;
  const _ZoneDef({required this.label, required this.pointKeys, this.chartKey, required this.actionsRelevant});
}

const zoneDefs = <String, _ZoneDef>{
  'living': _ZoneDef(
    label: 'Living',
    pointKeys: ['house.occupied', 'living.pir', 'living.radar'],
    actionsRelevant: false,
  ),
  'utility': _ZoneDef(
    label: 'Water / Utility',
    pointKeys: [
      'tank.line.flow_lpm', 'water_tank.level_pct', 'main_valve.valve_state',
      'pump.pump_state', 'pump.current_a', 'virtual.water.leak_suspected',
      'virtual.pump.dryrun_suspected', 'health.main_valve', 'health.pump',
    ],
    chartKey: 'water_tank.level_pct',
    actionsRelevant: true,
  ),
  'panel': _ZoneDef(
    label: 'Main Panel',
    pointKeys: ['main_panel.power_w'],
    chartKey: 'main_panel.power_w',
    actionsRelevant: false,
  ),
};

class RoomScreen extends StatelessWidget {
  final String id;
  const RoomScreen({super.key, required this.id});

  @override
  Widget build(BuildContext context) {
    final def = zoneDefs[id];
    final store = StoreScope.of(context);
    return Scaffold(
      appBar: AppBar(title: Text(def?.label ?? 'Unknown room')),
      body: def == null
          ? ListView(
              padding: const EdgeInsets.all(DomoraSpace.s4),
              children: [
                GlassCard(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      const CardTitle('Unknown room'),
                      EmptyNote('No zone named "$id". Try Living, Water / Utility, or Main Panel from Home.'),
                    ],
                  ),
                ),
              ],
            )
          : AnimatedBuilder(animation: store, builder: (context, _) => _body(context, store, def)),
    );
  }

  Widget _body(BuildContext context, DomoraStore store, _ZoneDef def) {
    final zones = deriveZones(store.points);
    final zone = zones.firstWhere((z) => z.label == def.label,
        orElse: () => Zone(id: id, label: def.label, state: DomoraState.ok, detail: '—'));
    final water = deriveWater(store.points);
    final occ = deriveOccupancy(store.points);
    final powerW = derivePowerW(store.points);

    return ListView(
      padding: const EdgeInsets.all(DomoraSpace.s4),
      children: [
        GlassCard(
          child: Row(
            children: [
              HealthDot(zone.state),
              const SizedBox(width: DomoraSpace.s3),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(def.label,
                        style: const TextStyle(fontSize: 20, fontWeight: FontWeight.w700, color: DomoraColors.ink)),
                    Text(zone.detail,
                        style: const TextStyle(fontFamily: 'monospace', fontSize: 12, color: DomoraColors.inkFaint)),
                  ],
                ),
              ),
            ],
          ),
        ),
        if (def.chartKey != null) ...[
          const SizedBox(height: DomoraSpace.s4),
          GlassCard(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                const CardTitle('Trend — this session'),
                DomoraLineChart(
                  data: store.seriesFor(def.chartKey!),
                  valueFmt: (v) => v.toStringAsFixed(1),
                ),
              ],
            ),
          ),
        ],
        if (id == 'living') ...[
          const SizedBox(height: DomoraSpace.s4),
          GlassCard(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                const CardTitle('Occupancy'),
                Wrap(
                  crossAxisAlignment: WrapCrossAlignment.center,
                  spacing: DomoraSpace.s2,
                  runSpacing: 4,
                  children: [
                    DomoraChip(occ.occupied ? 'occupied' : 'empty',
                        state: occ.occupied ? DomoraState.ok : DomoraState.learn),
                    Text('radar ${occ.radar ? "on" : "off"} · PIR ${occ.pir ? "on" : "off"}',
                        style: const TextStyle(fontFamily: 'monospace', fontSize: 12, color: DomoraColors.inkFaint)),
                  ],
                ),
              ],
            ),
          ),
        ],
        if (id == 'utility') ...[
          const SizedBox(height: DomoraSpace.s3),
          Wrap(
            spacing: DomoraSpace.s2,
            runSpacing: DomoraSpace.s2,
            children: [
              DomoraChip(water.leak ? 'leak suspected' : 'no leak',
                  state: water.leak ? DomoraState.crit : DomoraState.ok),
              DomoraChip(water.dryrun ? 'pump dry-run' : 'pump nominal',
                  state: water.dryrun ? DomoraState.crit : DomoraState.ok),
            ],
          ),
        ],
        if (id == 'panel' && !powerW.isNaN) ...[
          const SizedBox(height: DomoraSpace.s4),
          GlassCard(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                const CardTitle('Power now'),
                Text(fmtWatts(powerW),
                    style: const TextStyle(fontSize: 28, fontWeight: FontWeight.w700, color: DomoraColors.ink)),
              ],
            ),
          ),
        ],
        const SizedBox(height: DomoraSpace.s4),
        GlassCard(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              const CardTitle('Points in this room'),
              for (final key in def.pointKeys)
                _PointRow(
                  pointKey: key,
                  value: store.points.containsKey(key) ? fmtValue(store.points[key]!.value) : '—',
                  onTap: () => pushScoped(context, SensorScreen(pointKey: key)),
                ),
            ],
          ),
        ),
        const SizedBox(height: DomoraSpace.s4),
        GlassCard(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              CardTitle(def.actionsRelevant ? 'Autonomous actions in this room' : 'Autonomous actions'),
              if (!def.actionsRelevant)
                const EmptyNote(
                    'This system never acts on this room — only the water loop (utility) is autonomous today.')
              else if (store.actionOrder.isEmpty)
                const EmptyNote('No autonomous action yet in this run.')
              else
                for (final aid in store.actionOrder)
                  IncidentRow(action: store.actions[aid]!, detailed: false),
            ],
          ),
        ),
      ],
    );
  }
}

class _PointRow extends StatelessWidget {
  final String pointKey;
  final String value;
  final VoidCallback onTap;
  const _PointRow({required this.pointKey, required this.value, required this.onTap});

  @override
  Widget build(BuildContext context) => InkWell(
        onTap: onTap,
        borderRadius: BorderRadius.circular(DomoraRadius.r1),
        child: Padding(
          padding: const EdgeInsets.symmetric(vertical: 7),
          child: Row(
            children: [
              Expanded(
                flex: 3,
                child: Text(pointKey,
                    overflow: TextOverflow.ellipsis,
                    style: const TextStyle(fontFamily: 'monospace', fontSize: 12, color: DomoraColors.inkDim)),
              ),
              const SizedBox(width: DomoraSpace.s2),
              Expanded(
                flex: 2,
                child: Text(value,
                    overflow: TextOverflow.ellipsis,
                    textAlign: TextAlign.right,
                    style: const TextStyle(fontFamily: 'monospace', fontSize: 12, color: DomoraColors.ink)),
              ),
              const Icon(Icons.chevron_right_rounded, size: 16, color: DomoraColors.inkFaint),
            ],
          ),
        ),
      );
}
